provider/eventstore/eventstore.go

```diff
@@ -0,0 +1,636 @@
+	"github.com/inklabs/rangedb/pkg/broadcast"
+	"github.com/inklabs/rangedb/pkg/clock"
+	"github.com/inklabs/rangedb/pkg/clock/provider/systemclock"
+	"github.com/inklabs/rangedb/pkg/rangedberror"
+	"github.com/inklabs/rangedb/pkg/recordsubscriber"
+	"github.com/inklabs/rangedb/pkg/shortuuid"
+	"github.com/inklabs/rangedb/provider/jsonrecordserializer"
+)
+
+const (
+	rpcErrContextCanceled   = "Canceled desc = context canceled"
+	streamNotFound          = "Failed to perform read because the stream was not found"
+	broadcastRecordBuffSize = 100
+)
+
+type eventStore struct {
+	clock         clock.Clock
+	serializer    rangedb.RecordSerializer
+	uuidGenerator shortuuid.Generator
+	broadcaster   broadcast.Broadcaster
+	ipAddr        string
+	username      string
+	password      string
+
+	sync                 sync.RWMutex
+	version              int64
+	globalSequenceNumber uint64
+	savedStreams         map[string]struct{}
+}
+
+// Option defines functional option parameters for eventStore.
+type Option func(*eventStore)
+
+// WithClock is a function option to inject a clock.Clock
+func WithClock(clock clock.Clock) Option {
+	return func(store *eventStore) {
+		store.clock = clock
+	}
+}
+
+// WithUUIDGenerator is a functional option to inject a shortuuid.Generator.
+func WithUUIDGenerator(uuidGenerator shortuuid.Generator) Option {
+	return func(store *eventStore) {
+		store.uuidGenerator = uuidGenerator
+	}
+}
+
+// WithSerializer is a functional option to inject a rangedb.RecordSerializer.
+func WithSerializer(serializer rangedb.RecordSerializer) Option {
+	return func(store *eventStore) {
+		store.serializer = serializer
+	}
+}
+
+// New constructs an eventStore. Experimental: Use at your own risk.
+func New(ipAddr, username, password string, options ...Option) *eventStore {
+	s := &eventStore{
+		clock:        systemclock.New(),
+		serializer:   jsonrecordserializer.New(),
+		broadcaster:  broadcast.New(broadcastRecordBuffSize, broadcast.DefaultTimeout),
+		savedStreams: make(map[string]struct{}),
+		ipAddr:       ipAddr,
+		username:     username,
+		password:     password,
+	}
+
+	for _, option := range options {
+		option(s)
+	}
+
+	return s
+}
+
+func (s *eventStore) NewClient() (*esclient.Client, error) {
+	config, err := esclient.ParseConnectionString(fmt.Sprintf("esdb://%s:%s@%s", s.username, s.password, s.ipAddr))
+	if err != nil {
+		return nil, fmt.Errorf("unexpected configuration error: %s", err.Error())
+	}
+
+	config.DisableTLS = true
+	config.SkipCertificateVerification = true
+	client, err := esclient.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create client: %s", err.Error())
+	}
+
+	return client, nil
+}
+
+func (s *eventStore) streamName(name string) string {
+	s.sync.RLock()
+	defer s.sync.RUnlock()
+	return fmt.Sprintf("%d-%s", s.version, name)
+}
+
+func (s *eventStore) SetVersion(version int64) {
+	s.sync.Lock()
+	s.globalSequenceNumber = 0
+	s.version = version
+	s.sync.Unlock()
```

Comment:
Go question, why can't we do `defer s.sync.Unlock()` here?

Also, why do we need the lock?